New feature: Enriched signatures given a list of genes

dongbohu opened the pull request:
This PR is for issue #153. Part of the implementation is based on @rzelayafavila's PR #343.

dongbohu commented:
Here is a demo page:
http://165.123.67.202/#/gene_search?mlmodel=4
Type in "dna" in search box, then select "dnaA", now click "Find Enriched Signatures", the page should jump to:
http://165.123.67.202/#/signature/enriched_signatures?mlmodel=4&genes=6546

dongbohu commented on adage/analyze/api.py:
```diff
@@ -472,8 +472,30 @@ class Meta:
 
+    def apply_filters(self, request, applicable_filters):
+        """Implementation of "related_genes" filter, which allows the
+        API to get all pratitipation records of signatures that includes
```
Good catch. thanks!

mhuyck commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are genes names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
```
Thank you for the awesome comment explaining the structure of `groupedGenes` here! The variable isn't used until we get down to `getEnrichedSignatures()` at line 165 below. Could we move this comment and definition down there to minimize the scrolling needed to refer to this comment as a new reader tries to grok the code there?

dongbohu replied:
That's a good point. I will do that.

mhuyck commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+        var genes = genesBySignatures[sigID];
+        var n = Object.keys(genes).length, k = 0;
+        var matchedGenes = [], selectedGene;
+        var pValue;
```
I know this is a lot of `var`s to define, but could we please have just one per line? That makes it easier to scan and find where each one is defined. If you don't like writing `var` on each line, you could say:

```javascript
var genes = genesBySignatures[sigID],
  n = Object.keys(genes).length,
  k = 0,
  matchedGenes = [],
  selectedGene,
  pValue;
```

dongbohu replied:
Sure!

mhuyck commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+        var matchedGenes = [], selectedGene;
+        var pValue;
+        for (var i = 0; i < genesInUrl.length; i++) {
+          selectedGene = genesInUrl[i];
```
You had me convinced to prefer `.forEach()` in most places. I don't see why we shouldn't use that here... I think it would be easier to read:

```javascript
genesInUrl.forEach(function(selectedGene) {
...
});
```

dongbohu replied:
Good question. THat was what I wrote originally, but `jshint` bundled in our `Gruntfile.js` gave me an error when building. It doesn't allow a function defined in a loop. This rule is mainly designed to avoid a bug in such a program:
```
for (var i = 1; i <= 5; i++) {
  setTimeout(function() { console.log(i); }, i * 1000);
}
```
I think my current program doesn't have this bug, but to make jshint happy (and avoid any potential bug), I am using the traditional `for` loop.

mhuyck replied:
Ack! What a pain... Yes, I have definitely encountered this limitation in `jshint` as well. I guess the pragmatic solution is to leave the `for` loop for now, then. We are definitely going to need to take the time to update our build tools in the not too distant future.

mhuyck commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+        // m: the number of genes in genesInUrl;
+        // n: the total number of heavy genes related to sigID;
+        // N: the total number of genes.
+        pValue = 1 - MathFuncts.hyperGeometricTest(k, m, n, N);
```
I need to pull out my stats notes to try and follow this test more carefully. In the meantime, can you tell me how you came up with the inputs here? As I mentioned above, I am not sure the `N` that's going in is what we want. We should double-check the rest of the parameters, too. Ideally, we would have a test case to confirm that the output matches what we expect.

cgreene replied:
I had a conversation with @rzelayafavila about this. The ideal set of genes was the set of genes that _could have actually been returned_. This set contains the genes that were included in the model & that are also in our database. Obtaining that was a time consuming query (and it meant we would need to filter only the query genes to this same set).

IIRC @rzelayafavila checked into this and that number was essentially the same as the number of PA genes in our database. So I think that he decided to use just _the number of genes that exist in the DB for this organism_. This will be serviceable until we go to systems where those differ substantially. I think he was going to file an issue on this and/or add a comment to note the limitations of this approach and why he was selecting it at this time.

Of course, if you want to revisit this and/or you're not concerned about those filters, it'd be possible to do the original/ideal query.

dongbohu replied:
Rene's original PR includes the following comment:
```
    // For the moment, the total number of genes in our universe
    // to perform this enrichment analysis, is the total number of
    // Entrez IDs in our database for Pseudomonas aeruginosa.
```
I think this is probably the comments that @cgreene was referring to. (I replaced these lines with a shorter line because it seemed redundant.) I will add the comments back.

rzelayafavila replied:
Yes, exactly, that was the reasoning for getting the total number of genes from the DB for Pseudomonas.

mhuyck replied:
Okay, that makes way more sense with the comment then. How can we ensure that the query only returns genes for *pseudomonas*, then? Looking back at `enriched_signatures.js`, line 76, we request *all* genes from the database (without any filter).

I guess you're saying that our database *only* has genes for *pseudomonas*. At some point we plan to add other organisms... how will we remember to fix that line in the future when we begin work on that?

We need an issue, for sure, and we have a milestone called `Before multi-organism`, so the issue should get filed and tagged that way. I think I would take one additional step and add a line of code to our management command for loading genes that will make it halt with an error if we try to load genes for any organism other than *pseudomonas*...

...aha. We can't alter the management command because it is from `django-genes`. The next best thing is a comment in the scripts that call it. I see two places: `adage_server.py`, line 180, and `load_default_pseudomonas_data.sh`, line 40. The comments should make specific mention of the relevant issue.

mhuyck replied:
Another approach is to add a unit test that will fail (and halt deployment) if it finds genes from any organism other than *pseudomonas* (`taxonomy_id`=208964). The comments in the test should make specific reference to the issue.

mhuyck commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+      var correctedPValues = MathFuncts.multTest.fdr(pValueArray);
+      correctedPValues.forEach(function(element, index) {
+        var correctedPValue = element.toPrecision(pValueSigDigits);
+        if (correctedPValue < self.pValueCutoff) {
```
This test for the `pValueCutoff` should happen *before* the `.toPrecision()` is done. We should use all of the precision we have to make the cutoff determination, while the `.toPrecision(pValuesSigDigits)` can be thought of as "formatting" the output.

If we don't do so in the right order, rounding could bump some values above the cutoff inappropriately.

dongbohu replied:
Good point. I will do that.

dongbohu commented on interface/src/app/gene/searchFew/autocomplete-search-form.tpl.html:
```diff
@@ -35,6 +35,10 @@
     </button>
+    &nbsp;
+    <button class="btn btn-primary btn-lg" ng-click="findEnrichedSignatures()">
+      Find Enriched Signatures <i class="fa fa-bars"></i>
```
This file was copied from Rene's PR. I think what you said makes sense. I will correct the two buttons.

rzelayafavila replied:
Ah I see. Yes, I had originally put it in to refer to the fact that we were going to show a list/table, but that makes sense.

dongbohu commented on adage/analyze/api.py:
```diff
@@ -472,8 +472,30 @@ class Meta:
+
+            signatures = Participation.objects.filter(
+                gene__in=query_genes).values('signature').distinct()
+            object_list = object_list.filter(signature__in=signatures)
```
I tested the query above and it was slow. My first thought was to create indexes for `signature` and `gene` fields in `Participation` table, but then I realized that both are foreign keys, so they are already indexed. Due to the nature of this query, I don't think it can be done by going through the participation records just once. One thing I can think of is to get all participation records first from the database, then filter the records based on input `related_genes`.

dongbohu replied:
@mhuyck: I did a few more tests and found that the most time consuming part is actually this line:
```
object_list = object_list.filter(signature__in=signatures)
```
because `gene` field is a foreign key in `ParticipationResource`, the more records in the returned `object_list`, the more table joins will be needed to get the details of gene. Since `signature` and the current module do need the gene information, we can not change line 461 into:
```
gene = fields.IntegerField(attribute='gene_id', null=False)
```

mhuyck replied:
Rats. I thought this would be simpler to speed up. If it's not a quick thing to optimize the filter then we should log it as an issue and come back to it. It's better to *have* the feature and see how it gets used before spending a lot of effort to make it perfect.

cgreene replied:
☝️  Do we have a spinner we could easily have up as a filler while this happens so the user knows something is being done?

dongbohu replied:
Let me find a spinner.

dongbohu commented on interface/src/app/gene/searchFew/searchFew.js:
```diff
@@ -76,6 +76,10 @@ angular.module('adage.gene.searchFew', [
+
+        $scope.findEnrichedSignatures = function() {
+          CommonGeneFuncts.findEnrichedSignatures($scope, $state);
+        };
```
Let me create a new issue to keep track of it.

dongbohu commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+ * "adage.enrichedSignatures" module.
+ */
+
+angular.module('adage.enrichedSignatures', [
```
It probably makes more sense to move this file (and the html template) to `gene` directory, because it is based on a search of genes. What do you think? (I was thinking about it when reading Rene's PR but forgot later.)

mhuyck replied:
Ah! I see what you mean. It could definitely go in either place, since this feature examines both genes and signatures. I think about this in two ways: (1) the ideal component arrangement, and (2) pragmatically, what works best?

1. From a component-driven (or object-oriented) perspective, the model for `Gene` is more abstract than the model for `Signature`. There can be no `Signature` without the concept of a `Gene`, but a `Gene` does not depend upon the concept of a `Signature`. This indicates that the logic for finding enriched signatures ideally belongs in the component for `Signature`s. The implementation for a `Gene` should not "know" anything about a `Signature`.

2. From a pragmatic perspective, the code should be found easily by a new programmer (or one of us, a year from now) looking for how to change this feature. Since the page is called "enriched signatures," I'd say it would be easiest to find in the `app/signature` component.

So I think both perspectives point toward keeping it there.

dongbohu commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+    });
+
+    // Promise that gets all signatures of given mlmodel in URL:
+    var p1 = Signature.get(
```
Sure.

dongbohu commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+    var p1 = Signature.get(
+      {mlmodel: self.modelInUrl, limit: 0},
+      function success(response) {
+        self.signatures = Object.create(null);
```
I agree that `{}` seems more readable, but I can't resist the advantage that`Object.create()` offers. (I checked its compatibility too before using it in this file.)

dongbohu commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+    var p2 = Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
```
See the comments from @cgreene and me below.

dongbohu commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,203 @@
+      }
+    ).$promise;
+
+    var pValueSigDigits = 3;
```
Sure. This constant can be used in `signature.js` too, because it also calculates the p-value.

dongbohu commented:
I modified the code based on your comments. `enriched_signatures.js` has been revised a lot due to one issue that I just realized: the `get` request` of `Gene` should include organism ID.

rzelayafavila commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,216 @@
+
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
```
@dongbohu - I think this `k` parameter is correct, in that it is the number of genes that are both in the `genesInUrl` and have some kind of participation in the desired signature. However, are we only doing "heavy" genes or all types of participation? Should this be reflected in the comments? Also, the `n` parameter below also mentions "heavy genes".

rzelayafavila commented on interface/src/app/signature/enriched_signatures.js:
```diff
@@ -0,0 +1,216 @@
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
+        // m: the number of genes in genesInUrl;
```
Ah, I just thought of something for this parameter - do you think we need to check whether all of the genes in the url are found in any of the signatures for a specific model? This is because the `m` parameter is the number of "successes" in the population. Maybe each of the genes in the `genesInUrl` should only be considered a "success" if they are found somewhere in the population (meaning they have a participation tying them to one or more signatures for this model)?

rzelayafavila replied:
Hmmm, on second thought, I think this might be a moot point, since the `N` parameter below (which is the whole "population" of genes) takes into account ALL of the genes in the database for this organism (as we discussed), and the `genesInUrl` can only be drawn from that population of genes, right? So I think that as long as `N` is the total number of genes in the DB for that organism, this `m` parameter can just be the number of all genes in `genesInUrl`.

dongbohu commented:
@rzelayafavila Thank you for your comments. What you said about "heavy genes" in the comments definitely makes sense. About the calculation of `m`, I don't know whether we should filter out the genes in URL that don't have any participation record in the database.

dongbohu commented:
@mhuyck Before seeing your comments on the possible locations of `enriched_signatures.js` and `enriched_signatures.tpl.html`, I have moved both of them to `gene` directory and commited the change. I feel it makes (at least some) sense because this page is derived from `genes` page, and the enrichment calculation depends on the genes selected by users from genes web UI. What do you think?

I noticed the defects of calculation of `N` a few days ago and had corrected it. The latest change now includes two parameters when requesting genes from back end. One of them is the organism ID (which is collected from the ML model).

@cgreene: A spinner is also added when data is loading from server. See this demo page:
http://165.123.67.202/#/signature/enriched_signatures?mlmodel=4&genes=6546,6547,6548

mhuyck commented:
Oops! Well, I guess my preference for keeping the code in `app/signature` is pretty clear. I see your reasoning behind putting into `app/gene` as well, since it is an operation on collections of genes. Both you and @rzelayafavila have a deeper understanding of how the genes and signatures are organized, so let me know if there's a better way to think about it. I can reorganize my head, too, if the code should stay as it is.

mhuyck commented:
Good catch on adding an organism filter to the query that finds `N` from the back-end. The loading message is a nice touch.

dongbohu commented:
@mhuyck If you look at the files in `signature` directory, they are all derived from the main signature module, that is why I decided to put the new `enriched_signatures` files in `gene` directory.

dongbohu commented:
It may make even more sense to create a sub-directory `gene/enriched_signatures/` and put these two new files there. What do you think @mhuyck?

dongbohu commented:
@mhuyck I made a few extra changes, including a reminder message at the bottom of `gene-network.tpl.html` for users (see my discussion with Deb in `adage-team` channel on slack). I also added a new issue about the performance when computing gene-gene network and enriched signatures. 

The only thing that I haven't done is create a new directory `gene/enriched_signatures` to host the two files related to enriched signatures computation. What do you think?

mhuyck commented on interface/src/app/signature/signature.js:
```diff
@@ -342,8 +344,7 @@ angular.module('adage.signature', [
 
-        $scope.pValueCutoff = 0.05;
-        var pValueSigDigits = 3;
+        $scope.pValueCutoff = pValueCutoff;
```
👍

mhuyck commented on interface/src/app/signature/signature.js:
```diff
@@ -407,10 +408,9 @@ angular.module('adage.signature', [
-            if (correctedPValue < $scope.pValueCutoff) {
+            if (correctedPValues[i] < $scope.pValueCutoff) {
+              var correctedPValue = correctedPValues[i].toPrecision(
+                pValueDigits);
```
👍  Nice catch!

mhuyck commented:
Oh, I now see your question above about creating an `enriched_signatures` subdirectory. I see your point about organizing this way. Let's leave things as is for now. We can think about moving the file around later if it improves the overall organization of the source tree.

dongbohu commented on interface/src/app/gene/enriched_signatures.js:
```diff
@@ -0,0 +1,222 @@
+        // The only field we care in response is 'total_count', so
+        // 'limit' is set to 1 to save some bandwidth.
+        apiPromises.push(Gene.get(
+          {organism: $scope.mlmodel.organism.id, limit: 1},
```
Yes, I thought of that approach too. But I tried "limit=1" first and felt that it was fast enough, so I didn't bother further optimization.

dongbohu commented on interface/src/app/utils.js:
```diff
@@ -18,6 +18,13 @@ angular.module('adage.utils', [])
 .constant('ActivityDigits', 5)
 
+// p-value cutoff
+// Number of digits in p-value shown on frontend:
```
Good catch! Thanks.

dongbohu commented on interface/src/app/gene/enriched_signatures.js:
```diff
@@ -0,0 +1,222 @@
+      });
+
+      var participationTypes = Object.keys(groupedGenes);
+      participationTypes.forEach(function(typeName) {
```
Yes, that's right.

dongbohu commented on interface/src/app/gene/searchFew/searchFew.js:
```diff
@@ -76,6 +75,10 @@ angular.module('adage.gene.searchFew', [
         };
+
+        $scope.findEnrichedSignatures = function() {
+          CommonGeneFuncts.findEnrichedSignatures($scope, $state);
```
Not really, because `searchFew.js` includes `adage.gene.selected` as a dependency, which includes `adage.gene.utils` as a dependency. See line 2 of `gene/selected/selected.js` file.

mhuyck replied:
Aha... that explains it then. It still feels weird to have this file depend upon `CommonGeneFuncts` without declaring that dependency, but if it's implicitly included already then we've technically got it covered.

dongbohu commented on interface/src/app/gene/enriched_signatures.js:
```diff
@@ -0,0 +1,222 @@
+        }
+      });
+      self.statusMessage = '';
+      self.inProgress = false;
```
I like the idea of using `finally`. Actually I realized that the current error handling has a bug. Since each promise has its own error handling function, even if any of these promises is rejected, `getEnrichedSignatures` function will still be called. So I need to throw an exception inside each error handling function so that the error can be detected by `catch()` and `getEnrichedSignatures` will be skipped.

mhuyck replied:
Even better. That looks really great now! Much easier to follow what is going on, and less repeated code. Even the curmudgeonly Codeclimate seems happier about it.

dongbohu commented:
@mhuyck: I believe I have addressed all your comments. Please take a look again. Thanks!

mhuyck commented on interface/src/app/gene/enriched_signatures.js:
```diff
@@ -0,0 +1,222 @@
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
```
Issue logged as #349.

mhuyck commented:
Nice work, @dongbohu!

dongbohu commented:
@mhuyck Thank you very much for creating that issue. I should have done that yesterday.
